### PR TITLE
Add experimental support to write incoming data to a Kafka-compatible backend

### PR DESCRIPTION
**Pull request by pracucci**
#### What this PR does

We're building a prototype of an alternative Mimir architecture where the write and read path are fully decoupled, with a Kafka-compatible backend in between. This is going to be a multi-quarter effort and we would like to progressively upstream code changes while building it. The idea is that we'll do our best to keep these changes isolated from the rest of Mimir as much as possible, with few integration points.

In this PR I'm proposing to upstream a basic support to write incoming requests from distributor to a Kafka-compatible backend.

Notes:
- Configuration is hidden in the auto-generated documentation to avoid confusion to final users while we build the prototype.
- No CHANGELOG entry while building the prototype, for the same reason as above: to avoid confusion to final users.
- The plugging in the distributor is not the final one. Currently it's a very hacky way to partition data, while @pstibrany is working on the definitive partitioning scheme.
- This PR also introduces the franz-go Kafka client, which is the library we selected after testing the [segmentio](https://github.com/segmentio/kafka-go) one and this one. Based on our tests, the franz-go library looks more evoluted, mature and stable than the segmentio one.

#### Which issue(s) this PR fixes or relates to

N/A

#### Checklist

- [ ] Tests updated.
- [ ] Documentation added.
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/storage/ingest/writer_test.go [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+			require.Equal(t, kgo.ErrRecordTimeout, writer.WriteSync(ctx, partitionID, tenantID, series1, nil, mimirpb.API))
+			elapsedTime := time.Since(startTime)
+
+			// It should take nearly the client's write timeout.
+			expectedElapsedTime := writerCfg.KafkaWriteTimeout + writerRequestTimeoutOverhead
+			tolerance := time.Second
+			assert.Greater(t, elapsedTime, expectedElapsedTime-tolerance)
+			assert.Less(t, elapsedTime, expectedElapsedTime+tolerance)
+		})
+
```
Comment (dimitarvdimitrov) [inline pkg/storage/ingest/writer_test.go:294]: can use 
```go
			assert.InEpsilon(t, expectedElapsedTime, elapsedTime, float64(time.Second))
```

File: pkg/storage/ingest/writer.go [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+		kgo.DialTimeout(w.kafkaCfg.KafkaDialTimeout),
+		kgo.WithHooks(metrics),
+		kgo.WithLogger(newKafkaLogger(logger)),
+
+		// Use a static partitioner because we want to be in control of the partition.
+		kgo.RecordPartitioner(newKafkaStaticPartitioner(int(partitionID))),
+
+		// Set the upper bounds the size of a record batch. We intentionally don't set the producer linger
+		// because we expect the producer to be high volume and in such case the library doc recommend to not
+		// set linger.
```
[the remark this answers is not in the document]
Comment (pracucci) [reply pkg/storage/ingest/writer.go:227]: Oh right! I wanted to update it, then I forgot.

File: pkg/storage/ingest/writer_test.go (hunk 1 above)
Comment (pracucci) [reply pkg/storage/ingest/writer_test.go:303]: But I think we should use `InDelta()` given we pass the absolute max diff value.

File: pkg/storage/ingest/writer.go [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+	// The following settings can only be overridden in tests.
+	kafkaMaxInflightProduceRequests int
+}
+
+func (cfg *WriterConfig) RegisterFlags(f *flag.FlagSet) {
+	cfg.RegisterFlagsWithPrefix("", f)
+}
+
+func (cfg *WriterConfig) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
+	f.DurationVar(&cfg.KafkaWriteTimeout, prefix+".kafka-write-timeout", 10*time.Second, "How long to wait for a incoming write request to be successfully committed to the Kafka backend. The request's data will be written to Kafka even if the incoming write request is cancelled. When a write fails, other buffered requests will fail too.")
```
[the remark this answers is not in the document]
Comment (pracucci) [reply pkg/storage/ingest/writer.go:44]: > Does this refer to the remote-write request being cancelled?

Yes.

> Can't the write request be cancelled before it reaches the sendToStorage function?

Yes, right.

> does a write here refer to a kafka write? If yes, then I don't understand why this should be in the flag help text and not somewhere in the code

Yes, it refers to a write to Kafka.

Let me just remove this sentence from the CLI flag. I agree shouldn't be there.

File: pkg/storage/ingest/writer_test.go [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+		received := mimirpb.WriteRequest{}
+		require.NoError(t, received.Unmarshal(fetches.Records()[0].Value))
+		require.Len(t, received.Timeseries, len(multiSeries))
+
+		for idx, expected := range multiSeries {
+			assert.Equal(t, expected.Labels, received.Timeseries[idx].Labels)
+			assert.Equal(t, expected.Samples, received.Timeseries[idx].Samples)
+		}
+
```
[the remark this answers is not in the document]
Comment (pracucci) [reply pkg/storage/ingest/writer_test.go:92]: My take is: the byte-level interface is the Kafka client. Our Writer has a domain-level interface, which means we write and read domain-level data structures (so timeseries & co).

File: pkg/storage/ingest/writer.go [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+		// With the default config it would retry infinitely.
+		//
+		// Details of the involved timeouts:
+		// - RecordDeliveryTimeout: how long a Kafka client Produce() call can take for a given record. The overhead
+		//   timeout is NOT applied.
+		// - ProduceRequestTimeout: how a long to wait the response for the Produce request (the Kafka protocol message)
+		//   after being sent on the network. The actual timeout is increased by the configured overhead.
+		kgo.RecordRetries(math.MaxInt64),
+		kgo.RecordDeliveryTimeout(w.writerCfg.KafkaWriteTimeout),
+		kgo.ProduceRequestTimeout(w.writerCfg.KafkaWriteTimeout),
```
[the remark this answers is not in the document]
Comment (pracucci) [reply pkg/storage/ingest/writer.go]: _Discussed offline._

The TL;DR is that you depends on the actual failure scenario.

If the backend is slow, then an higher `ProduceRequestTimeout` increases the chances of a successful request (in other words, one try of 10s is better than 2x5s tries because maybe just waiting longer fixes it).

If the backend is unhealthy, then a shorter `ProduceRequestTimeout` is better because it will retry within the `RecordDeliveryTimeout` but the retry will be successful only if the meanwhile the cluster metadata has been updated and the replica owning a given partition has been moved to another broker, otherwise the client will just keep trying to connect to the previous (unhealthy) one. In a setup where these timeouts are relatively low, it may not be that common that the unhealthy replica has been actually detected as unhealthy and the replica owner for a given partition moved.

File: pkg/storage/ingest/config.go [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+var (
+	ErrMissingKafkaAddress = errors.New("the Kafka address has not been configured")
+	ErrMissingKafkaTopic   = errors.New("the Kafka topic has not been configured")
+)
+
+type Config struct {
+	Enabled bool `yaml:"enabled"`
+
+	KafkaConfig  KafkaConfig  `yaml:",inline"`
+	WriterConfig WriterConfig `yaml:",inline"`
```
[the remark this answers is not in the document]
Comment (pracucci) [reply pkg/storage/ingest/config.go]: Please take a look at the change in this commit: https://github.com/grafana/mimir/pull/6888/commits/ab77d27a8457f704b4869437d78f531fd0c85cb3

I've temporarily enabled the auto-generated doc and that's what you would get:

```diff
+ingest_storage:
+  # True to enable the ingestion via object storage.
+  # CLI flag: -ingest-storage.enabled
+  [enabled: <boolean> | default = false]
+
+  kafka:
+    # The Kafka backend address.
+    # CLI flag: -ingest-storage.kafka.address
+    [address: <string> | default = ""]
+
+    # The Kafka topic name.
+    # CLI flag: -ingest-storage.kafka.topic
+    [topic: <string> | default = ""]
+
+    # The Kafka client ID.
+    # CLI flag: -ingest-storage.kafka.client-id
+    [client_id: <string> | default = ""]
+
+    # The maximum time allowed to open a connection to a Kafka broker.
+    # CLI flag: -ingest-storage.kafka.dial-timeout
+    [dial_timeout: <duration> | default = 2s]
+
+    # How long to wait for an incoming write request to be successfully
+    # committed to the Kafka backend.
+    # CLI flag: -ingest-storage.kafka.write-timeout
+    [write_timeout: <duration> | default = 10s]
```

[the remark this answers is not in the document]
Comment (dimitarvdimitrov) [reply pkg/storage/ingest/config.go]: looks good, thanks for doing the change

**Review by dimitarvdimitrov — APPROVED** [review]
LGTM, thanks for addressing my comments

File: pkg/storage/ingest/writer.go [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+		// (which we don't need), we can increase the max number of in-flight Produce requests per broker. A higher
+		// number of in-flight requests, in addition to short buffering ("linger") in client side before firing the
+		// next Produce request allows us to reduce the end-to-end latency.
+		//
+		// The result of the multiplication of producer linger and max in-flight requests should match the maximum
+		// Produce latency expected by the Kafka backend in a steady state. For example, 50ms * 20 requests = 1s,
+		// which means the Kafka client will keep issuing a Produce request every 50ms as far as the Kafka backend
+		// doesn't take longer than 1s to process them (if it takes longer, the client will buffer data and stop
+		// issuing new Produce requests until some previous ones complete).
+		kgo.DisableIdempotentWrite(),
```
Comment (twmb) [inline pkg/storage/ingest/writer.go:237]: I think this is true in other clients, but in franz-go, disabling idempotency forces the client to only issue one request at a time -- franz-go favors not duplicating data. _With_ idempotency, franz-go allows 5 requests per broker (technically 4 due to some internal accounting but it's close enough).

Comment (twmb) [reply pkg/storage/ingest/writer.go:237]: Oh wait I missed MaxProduceRequestsInflightPerBroker just two lines down, my mistake. Nice!

Comment (pracucci) [reply pkg/storage/ingest/writer.go:237]: Thanks for looking at it and your feedback!
